clang/lib/CodeGen/CGRecordLayoutBuilder.cpp

```diff
@@ -394,33 +412,155 @@ void CGRecordLowering::accumulateFields() {
+      EndOffset = Offset;
+    }
+  }
+
+public:
+  RecordDecl::field_iterator begin() const { return Begin; }
+  RecordDecl::field_iterator end() const { return End; }
+
+public:
+  // Accessors
+  CharUnits getSize() const { return EndOffset - StartOffset; }
+  CharUnits getStartOffset() const { return StartOffset; }
+  CharUnits getEndOffset() const { return EndOffset; }
+
+  // Predicates
+  bool isBarrier() const { return getSize().isZero(); }
+  bool hasVolatile() const { return ContainsVolatile; }
+
+  // Create the containing access unit and install the bitfields.
+  void installUnit(CGRecordLowering &CGRL) const {
+    if (!isBarrier()) {
+      // Add the storage member for the access unit to the record. The
+      // bitfields get the offset of their storage but come afterward and remain
+      // there after a stable sort.
+      llvm::Type *Type = CGRL.getIntNType(CGRL.Context.toBits(getSize()));
+      CGRL.Members.push_back(CGRL.StorageInfo(getStartOffset(), Type));
+      for (auto F : *this)
+        if (!F->isZeroLengthBitField(CGRL.Context))
+          CGRL.Members.push_back(CGRecordLowering::MemberInfo(
+              getStartOffset(), CGRecordLowering::MemberInfo::Field, nullptr,
+              F));
+    }
+  }
+};
+
+// Create an access unit of contiguous bitfields.
+BitFieldAccessUnit::BitFieldAccessUnit(RecordDecl::field_iterator FieldBegin,
+                                       RecordDecl::field_iterator FieldEnd,
+                                       const CGRecordLowering &CGRL)
+    : BitFieldAccessUnit(FieldBegin, CharUnits::Zero(),
+                         FieldBegin->getType().isVolatileQualified()) {
+  assert(End != FieldEnd);
+
+  uint64_t StartBit = CGRL.getFieldBitOffset(*FieldBegin);
+  uint64_t BitSize = End->getBitWidthValue(CGRL.Context);
+  unsigned CharBits = CGRL.Context.getCharWidth();
+
+  assert(!(StartBit % CharBits) && "Not at start of char");
+
+  ++End;
+  if (BitSize ||
+      !(CGRL.Context.getTargetInfo().useZeroLengthBitfieldAlignment() ||
+        CGRL.Context.getTargetInfo().useBitFieldTypeAlignment()))
+    // The first field is not a (zero-width) barrier. Collect contiguous fields.
+    for (; End != FieldEnd; ++End) {
+      uint64_t BitOffset = CGRL.getFieldBitOffset(*End);
+      if (End->isZeroLengthBitField(CGRL.Context)) {
+        // A zero-length bitfield might be a barrier between access units.
+        if (CGRL.Context.getTargetInfo().useZeroLengthBitfieldAlignment() ||
+            CGRL.Context.getTargetInfo().useBitFieldTypeAlignment()) {
+          // A zero-length barrier, will be in the next run.
+          assert(!(BitOffset % CharBits) &&
+                 "Barrier bitfield not at byte boundary");
+          break;
+        }
+        // When this isn't a barrier, we want it to be placed at the end of the
+        // current run so that a final zero-length bitfield (why would one do
+        // that?) doesn't get mistaken for a barrier in later processing.
```

Comment:
We should honor zero-width bit-fields as barriers even on these terrible targets unless we don't have a choice.  That is, as long as the ABI hasn't actually colocated two bit-fields on opposite sides of a zero-width bit-field within the same byte, we should honor the zero-width bit-field and not let accesses interfere across it.

I think that might simplify the basic operation here — you want to accumulate bit-fields until the next bit-field starts at a byte boundary, and on well-behaved ABIs that means you'll never cross a zero-width bit-field, and on poorly-behaved ABIs you'll only cross the zero-width bit-fields that you absolutely have to.

I continue to feel like there's a simpler structure to this code that's trying to get out.  Can you think of this as doing a single pass over the bit-field run, accumulating into the current access unit, and then calling out to manage various "events" like reaching a barrier?  So basically:

```
func handleBitFieldRun() {
   var accumulator = {}
   var curAccessUnit = {}
   for field in run {
     if field.offset % byteSize {
       curAccessUnit.add(field)
     } else {
       if !curAccessUnit.empty() {
         accumulator.add(curAccessUnit)
         curAccessUnit = {}
       }
       if field.isZeroWidth {
         accumulator.barrier()
       } else {
         curAccessUnit.add(field)
       }
     }
   }
   accumulator.barrier()
}

func accumulator::barrier() {
  // emit the current access unit if we have one, then reset 
}

func accumulator::add(newAccessUnit) {
  // either merge the new access unit into the current one or
  // emit the current one and start building from the new one
}
```

Your merge algorithm is still online, right?  That is, you can think of it as always considering whether to merge the current access unit with the next, as opposed to deciding to merge the 2nd and 3rd and then later deciding whether to merge that with the 1st?

Reply:
As far as I know, zero-length bitfields always ensure that bitfields after the zero-length field don't overlap in the same byte as bitfields before the zero-length field.  The cross-target variations are related to other details which aren't relevant here, like the overall alignment of the struct.

Reply:
> As far as I know, zero-length bitfields always ensure that bitfields after the zero-length field don't overlap in the same byte as bitfields before the zero-length field. The cross-target variations are related to other details which aren't relevant here, like the overall alignment of the struct.

Unfortunately, the combination tested for here — where neither `useZeroLengthBitfieldAlignment()` nor `useBitFieldTypeAlignment()` is set — does cause struct layout to not even byte-align after a zero-width bit-field.  There are no targets that clear both bits by default, but one of them can be cleared with a flag, and I do know that we have some code at Apple that builds in that mode.  It's possible that we can break compatibility, but that's unconfirmed, so for now we should maintain existing behavior.

Reply:
Updated tests to include a set of {aarch64,arm}-apple-darwin targets.  If one is still missing please let me know